### PR TITLE
Fixed work secure_allocator on old compiers

### PR DESCRIPTION
**Pull request by TsynkPavel**
Not all compilers implement memset_s.
The proposed approach does the same.

issue: #1399

### CODE REVIEW

File: include/json/allocator.h [hunk 1]
```diff
@@ -38,8 +45,12 @@ template <typename T> class SecureAllocator {
    * The memory block is filled with zeroes before being released.
    */
   void deallocate(pointer p, size_type n) {
-    // memset_s is used because memset may be optimized away by the compiler
-    memset_s(p, n * sizeof(T), 0, n * sizeof(T));
+    #if defined(_MSC_VER)
+      RtlSecireZeroMemory(p, n * sizeof(T));
+    #else
+      for (auto it = reinterpret_cast<volatile unsigned char* volatile>(p); n; ++it, --n)
+        *it = 0;
```
Comment (tfc) [inline include/json/allocator.h:52]: It's just a nitpick, but why not:

```cpp
auto begin = reinterpret_cast<volatile unsigned char* volatile>(p);
std::copy(begin, begin + n, 0);
```

It's a little bit lighter on the eye while reading and will evaluate to nearly the same loop.

Comment (ZinkPavel) [reply include/json/allocator.h:52]: std::fill is available since c++17 this won't work for everyone

Comment (BillyDonahue) [reply include/json/allocator.h:52]: > std::fill is available since c++17 this won't work for everyone

not true. It's an old function

Comment (tfc) [reply include/json/allocator.h:52]: didn't mean to over-engineer this single line, but `std::fill_n(begin, n, 0);` is pretty declarative and available even in c++11.

Comment (ZinkPavel) [reply include/json/allocator.h]: Fixed it !

File: include/json/allocator.h [hunk 2]
```diff
@@ -38,8 +45,12 @@ template <typename T> class SecureAllocator {
    * The memory block is filled with zeroes before being released.
    */
   void deallocate(pointer p, size_type n) {
-    // memset_s is used because memset may be optimized away by the compiler
-    memset_s(p, n * sizeof(T), 0, n * sizeof(T));
+    #if defined(_MSC_VER)
+      RtlSecireZeroMemory(p, n * sizeof(T));
```
[the remark this answers is not in the document]
Comment (ZinkPavel) [reply include/json/allocator.h]: Fixed it !

File: include/json/allocator.h [hunk 3]
```diff
@@ -9,6 +9,13 @@
 #include <cstring>
 #include <memory>
 
+#if defined(_MSC_VER)
+  #if !defined(_WINSOCKAPI_)
```
[the remark this answers is not in the document]
Comment (ZinkPavel) [reply include/json/allocator.h]: Fixed it !

File: include/json/allocator.h [hunk 4]
```diff
@@ -38,8 +45,12 @@ template <typename T> class SecureAllocator {
    * The memory block is filled with zeroes before being released.
    */
   void deallocate(pointer p, size_type n) {
-    // memset_s is used because memset may be optimized away by the compiler
-    memset_s(p, n * sizeof(T), 0, n * sizeof(T));
+    #if defined(_MSC_VER)
```
[the remark this answers is not in the document]
Comment (ZinkPavel) [reply include/json/allocator.h]: Fixed it ! Add check memset_s to CMakeLists.txt

**Pull request merged**